gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java

```diff
@@ -149,12 +118,124 @@ public abstract static class Builder {
+
+    public abstract Builder setResolvedEndpoint(String resolvedEndpoint);
+
+    public abstract Builder setResolvedUniverseDomain(String resolvedUniverseDomain);
+
+    abstract String serviceName();
+
+    abstract String universeDomain();
+
+    abstract String clientSettingsEndpoint();
+
+    abstract String transportChannelProviderEndpoint();
+
+    abstract String mtlsEndpoint();
+
+    abstract boolean switchToMtlsEndpointAllowed();
+
+    abstract MtlsProvider mtlsProvider();
+
+    abstract boolean usingGDCH();
+
+    abstract String resolvedUniverseDomain();
+
     abstract EndpointContext autoBuild();
 
+    private String determineUniverseDomain() {
+      // Do not set the universe domain for GDC-H
+      if (usingGDCH()) {
+        return null;
+      }
+      // Check for "" (empty string)
+      if (universeDomain() != null && universeDomain().isEmpty()) {
+        throw new IllegalArgumentException("The universe domain value cannot be empty.");
+      }
+      // Override with user set universe domain if provided
+      return universeDomain() != null ? universeDomain() : GOOGLE_DEFAULT_UNIVERSE;
+    }
+
+    /** Determines the fully resolved endpoint and universe domain values */
+    private String determineEndpoint() throws IOException {
+      MtlsProvider mtlsProvider = mtlsProvider() == null ? new MtlsProvider() : mtlsProvider();
+      // TransportChannelProvider's endpoint will override the ClientSettings' endpoint
+      String customEndpoint =
+          transportChannelProviderEndpoint() == null
+              ? clientSettingsEndpoint()
+              : transportChannelProviderEndpoint();
+
+      // GDC-H has a separate flow
+      if (usingGDCH()) {
+        return determineGDCHEndpoint(customEndpoint);
+      }
+
+      // If user does not provide a custom endpoint, build one with the universe domain
+      if (Strings.isNullOrEmpty(customEndpoint)) {
+        customEndpoint = buildEndpointTemplate(serviceName(), resolvedUniverseDomain());
+      }
+
+      String endpoint =
+          mtlsEndpointResolver(
+              customEndpoint, mtlsEndpoint(), switchToMtlsEndpointAllowed(), mtlsProvider);
+
+      // Check if mTLS is configured with non-GDU
+      if (endpoint.equals(mtlsEndpoint())
+          && !resolvedUniverseDomain().equals(GOOGLE_DEFAULT_UNIVERSE)) {
+        throw new IllegalArgumentException(
+            "mTLS is not supported in any universe other than googleapis.com");
+      }
+
+      return endpoint;
+    }
+
+    // GDC-H has no concept of Universe Domain. Do not set the resolvedUniverseDomain value
+    private String determineGDCHEndpoint(String customEndpoint) {
+      if (universeDomain() != null) {
+        throw new IllegalArgumentException(
+            "Universe domain configuration is incompatible with GDC-H");
+      } else if (customEndpoint == null) {
+        return buildEndpointTemplate(serviceName(), GOOGLE_DEFAULT_UNIVERSE);
```

Comment:
Can we use `resolvedUniverseDomain` if we default it to GDU?

Reply:
Updated. Thanks!